### PR TITLE
Added types options to DeleteByQueryRequest

### PR DESCRIPTION
**Pull request by kunal642**
related to issue #21984

### CODE REVIEW

**Comment by nik9000** [comment]
I admit these requests in java are quite messy. Could you call `getSearchRequest.types(...)` instead?

**Comment by kunal642** [comment]
@nik9000 i did not get you
where do you want this call?

**Comment by kunal642** [comment]
@nik9000 from what i get from you comment, you do not want to add the types method and the user should use getSearchRequest.types(...) instead?

**Comment by nik9000** [comment]
> @nik9000 what i get from your comment is that you do not want to add the types method and the user should use getSearchRequest.types(...) instead
> am i correct?

Yes. I'm just sure it is worth adding this because that is an option.

**Comment by kunal642** [comment]
@nik9000 then the same should be followed for indices also.
getSearchRequest.indices(...) can be used to add the indices to the request and the indices method should also be removed from DeleteByQueryRequest.

**Comment by tlrx** [comment]
> getSearchRequest.indices(...) can be used to add the indices to the request and the indices method should also be removed from DeleteByQueryRequest.

That would make sense I agree, but these methods come from `IndicesRequest` and must be implemented by DeleteByQueryRequest so that request execution logic knows which indices are involved in the delete by query.

**Comment by kunal642** [comment]
@tlrx The IndicesRequest can be removed because only the indices methods are present in that interface and the DeleteByQueryRequest class does not need those methods.
the SearchRequest which is passed to DeleteByQueryRequest can have the indices and types specified in it.

File: core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java [hunk 1]
```diff
@@ -74,4 +75,23 @@ protected void extraRandomizationForSlice(DeleteByQueryRequest original) {
     protected void extraForSliceAssertions(DeleteByQueryRequest original, DeleteByQueryRequest forSliced) {
         // No extra assertions needed
     }
+
+    @Test
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java]: the Test annotation is banned in our project :) Something should fail when running the build from gradle. if the test method starts with test we are good, no need for the annotation

**Comment by javanna** [comment]
hi @kunal642 , the DeleteByQueryRequest needs to implement IndicesRequest as plugins may want to retrieve/replace the indices that the request applies to. IndicesRequest allows to do it in a generic manner.

**Comment by kunal642** [comment]
@javanna removed the annotation

**Comment by kunal642** [comment]
@javanna is there anything else that needs changes?

File: core/src/main/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequest.java [hunk 2]
```diff
@@ -111,4 +111,16 @@ public IndicesOptions indicesOptions() {
         assert getSearchRequest() != null;
         return getSearchRequest().indicesOptions();
     }
+
+    public String[] types() {
+        assert getSearchRequest() != null;
+        return getSearchRequest().types();
+    }
+
+    public IndicesRequest types(String...types) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequest.java]: can you return DeleteByQueryRequest here please? Also add a whitespace between `String...` and `types`?

**Review by javanna — CHANGES_REQUESTED** [review]
hi @kunal642 I had another look, left one comment, looks good otherwise

**Comment by kunal642** [comment]
@javanna the requested changes have been done

**Comment by kunal642** [comment]
@javanna Review this please

File: core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java [hunk 3]
```diff
@@ -74,4 +72,22 @@ protected void extraRandomizationForSlice(DeleteByQueryRequest original) {
...
+
+    public void testToDeleteByTypes() {
+        int numTypes = between(1, 50);
+        int numIndices = between(1, 100);
+        String[] indices = new String[numIndices];
+        String[] types = new String[numTypes];
+
+        for (int i = 0; i < numIndices; i++) {
+            indices[i] = randomSimpleString(random(), 1, 30);
+            types[i] = randomSimpleString(random(), 1, 30);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java]: This throws ArrayIndexOutOfBoundsException , can you fix that please?

File: core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java [hunk 4]
```diff
@@ -74,4 +72,22 @@ protected void extraRandomizationForSlice(DeleteByQueryRequest original) {
     protected void extraForSliceAssertions(DeleteByQueryRequest original, DeleteByQueryRequest forSliced) {
         // No extra assertions needed
     }
+
+    public void testToDeleteByTypes() {
+        int numTypes = between(1, 50);
+        int numIndices = between(1, 100);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java]: do we need to test indices as well as part of this test? Maybe only test types?

File: core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java [hunk 5]
```diff
@@ -74,4 +72,22 @@ protected void extraRandomizationForSlice(DeleteByQueryRequest original) {
     protected void extraForSliceAssertions(DeleteByQueryRequest original, DeleteByQueryRequest forSliced) {
         // No extra assertions needed
     }
+
+    public void testToDeleteByTypes() {
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequestTests.java]: can you please rename to `testTypesSetter` or something along those lines ?

**Review by javanna — CHANGES_REQUESTED** [review]
hi @kunal642 I left a couple of comments

**Comment by kunal642** [comment]
@javanna the requested changes have been done.

**Review by javanna — APPROVED** [review]
thanks @kunal642 I will merge this.

File: core/src/main/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequest.java [hunk 6]
```diff
@@ -111,4 +111,16 @@ public IndicesOptions indicesOptions() {
         assert getSearchRequest() != null;
         return getSearchRequest().indicesOptions();
     }
+
+    public String[] types() {
+        assert getSearchRequest() != null;
+        return getSearchRequest().types();
+    }
+
+    public DeleteByQueryRequest types(String... types) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/bulk/byscroll/DeleteByQueryRequest.java]: sorry @kunal642 one last thing, could you also test this method please? set the types using this new method and check that the underlying search request types reflect the change?

**Comment by kunal642** [comment]
@javanna done

**Pull request merged**